contracts/HomoraIBSwap.sol

```diff
@@ -0,0 +1,220 @@
+            path[1] = WETH;
+            path[2] = underlyingTokenOut;
+        }
+    }
+
+    /// @notice get the estimated outpout amount of each consecutive
+    /// swap along a path
+    /// @param tokenIn the address of the input token
+    /// @param tokenOut the address of the output token
+    /// @param amountIn the inputted amount of tokenIn
+    function getEstimatedAmountsOut(
+        address tokenIn,
+        address tokenOut,
+        uint256 amountIn
+    ) public view returns (uint256[] memory) {
+        return
+            uniswapRouter.getAmountsOut(amountIn, getPath(tokenIn, tokenOut));
+    }
+
+    /// @notice convert an amount of ibToken to that of the underlying token
+    /// @param ibToken the ibToken to get the underlying token amount of
+    /// @param ibTokenAmount the amount of ibToken to use for conversion
+    function ibToToken(address ibToken, uint256 ibTokenAmount)
+        external
+        view
+        returns (uint256)
+    {
+        CYToken cyToken = CYToken(SafeBox(ibToken).cToken());
+        uint256 tokenAmount =
+            ibTokenAmount.mul(cyToken.exchangeRateStored()).div(1e18);
+        return tokenAmount;
+    }
+
+    /// @notice convert an amount of token to that of the associated ibToken
+    /// @param ibToken the ibToken to get the token amount of
+    /// @param tokenAmount the amount of token to use for conversion
+    function tokenToIB(address ibToken, uint256 tokenAmount)
+        external
+        view
+        returns (uint256)
+    {
+        CYToken cyToken = CYToken(SafeBox(ibToken).cToken());
+        uint256 ibTokenAmount =
+            tokenAmount.mul(1e18).div(cyToken.exchangeRateStored());
+        return ibTokenAmount;
+    }
+
+    /// @notice swap an ibToken to another ibToken
+    /// @param tokenIn the input ibToken
+    /// @param tokenOut the desired output ibToken
+    /// @param amountIn the amount of tokenIn to swap
+    /// @param amountOutMin minimum amount of output tokens that must be received
+    /// for the transaction not to revert
+    /// @param deadline timestamp after which the transaction will revert
+    function swap(
+        address tokenIn,
+        address tokenOut,
+        uint256 amountIn,
+        uint256 amountOutMin,
+        uint256 deadline
+    ) external returns (uint256) {
+        require(tokenIn != tokenOut, "token-in-out-identical");
+        require(isIBToken[tokenIn], "token-in-not-supported");
+        require(isIBToken[tokenOut], "token-out-not-supported");
+
+        SafeBox safeboxIn = SafeBox(tokenIn);
+        safeboxIn.transferFrom(msg.sender, address(this), amountIn);
+        safeboxIn.withdraw(amountIn);
+
+        uint256 underlyingBalance;
+        address[] memory path = getPath(tokenIn, tokenOut);
+        uint256 outputAmount;
+
+        if (tokenIn != IBETHV2) {
+            IERC20 underlying = IERC20(safeboxIn.uToken());
+            underlyingBalance = underlying.balanceOf(address(this));
```

Comment:
For clarity, I think we should do `underlyingBalance = address(this).balance` in else case